### ISSUE TITLE
Top of progress bar not behaving as such

### ISSUE THREAD

**Issue by decodism** [iina/iina#3911]
**System and IINA version:**

- macOS 13.0 beta
- IINA 1.3.0

**Expected behavior:**

Clicking on the progress bar (on the height of its handle) should change the playback time.

**Actual behavior:**

Clicking on the top of the progress bar (on the height of its handle) does not change the playback time and dragging while holding down moves the control panel.

**Steps to reproduce:**

- Click on the top of the progress bar

**How often does this happen?**

Always

**Comment by decodism** [comment]
@low-batt maybe you could also look at this?

**Comment by low-batt** [comment]
I believe this screenshot shows the issue being reported:
![issue-3911-seek](https://user-images.githubusercontent.com/86170219/184557404-383d11fc-265c-454e-851e-2e43ec840e3c.jpg)

Notice the pointer is below the height of the seek bar's knob. Given that position the user would expect that clicking would seek playback to that position. Instead it initiates repositioning of the on screen controller (OSC).

The above screenshot was taken under Monterey 12.4. The same problem can be seen under Catalina 10.15.7, but not quite as bad. [NSSlider](https://developer.apple.com/documentation/appkit/nsslider) was changed by Apple for Big Sur. That may have made this problem slightly worse. I'm assuming Ventura behavior matches Monterey.

Another concern seen in the screenshot is that the time label overlaps the top of the seek bar's knob.

I will look into this.

**Comment by decodism** [comment]
Thank you for investigating this issue.

**Comment by decodism** [comment]
> Another concern seen in the screenshot is that the time label overlaps the top of the seek bar's knob.

That was the issue. The label was intercepting the clicks. Sending it to the back fixes the issue.

**Comment by low-batt** [comment]
That fully exposes the slider, but then the slider overlaps part of the time.

I'm close to posting a full analysis explaining what went wrong along with a more involved fix.
I might be able to finish it and post it tomorrow.

This is an important issue. Thank you for taking the time to report it!

**Comment by low-batt** [comment]
I have labeled this issue as a bug because the code is not working as intended.
I have labeled this issue as a high priority because it is a regression in macOS 11+.

### Analysis

Changing the `NSSlider` to have a green background makes it easier to see that the `NSTextField` showing the time intrudes into the slider when running under macOS 12.4:
![issue-3911-green](https://user-images.githubusercontent.com/86170219/185295031-7e7201d7-179a-41e1-82c8-b9113b6c86e5.jpg)


Instead of changing the slider the mouse down event results in the method `ControlBarView.mouseDown` being called which initiates dragging of the control bar.

Running under macOS 10.15.7 the situation is better. The `NSTextField` for the time just barely intrudes into the slider:
![issue-3911-catalina-green](https://user-images.githubusercontent.com/86170219/185295090-5052f206-ea29-448c-916e-ff44d833f63d.jpg)

Notice the slider seems taller when running under Monterey. Debug logging shows the slider's height as 23 points under Monterey and 19 points under Catalina.

I suspect the difference in behavior comes from changes Apple made to `NSSlider` in macOS 11. From [AppKit Release Notes for macOS Big Sur 11](https://developer.apple.com/documentation/macos-release-notes/appkit-release-notes-for-macos-11): 

> ### NSSlider
>
>  - The frame of [NSSlider](https://developer.apple.com/documentation/appkit/nsslider) has changed in macOS Big Sur 11.
>  - To ensure horizontal sliders appropriately align vertically with surrounding or nearby UI elements (like a label), use Auto Layout with the slider’s baseline.
>  - If your app cannot or does not use Auto Layout, you may query the [lastBaselineOffsetFromBottom](https://developer.apple.com/documentation/appkit/nsview/1525942-lastbaselineoffsetfrombottom) property to manually offset the view.
> - Some NSSliders you create using Interface Builder with Xcode 11 or earlier and that you open using Interface Builder with Xcode 12 won’t automatically have adjusted frames that match new control metrics in macOS Big Sur 11. These NSSliders may also appear visually clipped. To manually adjust the slider’s frame to match the new control metrics, either add Auto Layout constraints to the frame, or select the frame and and choose Editor > Size to Fit Content.

The Apple video [Adopt the new look of macOS](https://developer.apple.com/videos/play/wwdc2020/10104/) very briefly talks about the changes to `NSSlider`.

The Free Pascal Wiki page [macOS Big Sur changes for developers](https://wiki.lazarus.freepascal.org/macOS_Big_Sur_changes_for_developers#Controls) contains this table:

| Slider |	Size |	macOS 11 (default) |	macOS 10.15 and earlier (default) |
| --- | --- | --- | --- |
| Without tick marks |	Regular |	28pt |	19pt |
| |	Small |	20pt |	15pt |

Checking the slider in Xcode shows the `controlSize` is set to `regular` with a height of 28 points.

The logged height of 19 points under Catalina matches the size shown in the table above. But the logged height of 23 points under Monterey does not match up. The slider is in a view that is 29 points high with top and bottom layout constraints of 4.5 points each. There is enough room if the slider is 19 points high, but it appears the slider under Monterey is being compressed due to layout constraints. At least I think this is what is happening.

So the slider is bigger under Monterey, but why doesn't that result in the `NSTextField` holding the time being located higher in the OSC? Shouldn't the time preview be located relative to the top of the slider?

The lack of such adjustment is due to a use of a hardcoded [Magic Number](https://en.wikipedia.org/wiki/Magic_number_(programming)) in `MainWindowController`. As can be seen in this code from the method `updateTimeLabel` the time is being located 16 points from the base of the slider:

```swift
let timeLabelXPos = playSlider.frame.origin.y + 15
timePreviewWhenSeek.frame.origin = NSPoint(x: round(mouseXPos + playSlider.frame.origin.x - timePreviewWhenSeek.frame.width / 2),
                                           y: timeLabelXPos + 1)
```

The `15` and then plus `1` is interesting. Could it be that the slider's size was set to `small` (15) at the time this was coded? With the slider's height being 19 points in Catalina, 16 points puts the time `NSTextField` holding the time 3 points into the slider. Under Monterey the larger slider results in the time intruding 7 points into the slider.

### Fixing

The fix in the pull request restores what I think was the way the on screen controller was originally intended to behave. Another choice would be to adjust the OSC to support a larger progress bar slider in accordance with Apple's changes for Big Sur. That would be a more extensive change. If the OSC is to be revamped then the changes in PR #3529 for issue #548 should be included as they pertain to this area of the OSC.

The changes in the PR will eliminate the hardcoded numbers in the `MainWindowController` methods `playSliderChanges` and `updateTimeLabel` as seen here:
```swift
let timeLabelXPos = playSlider.frame.origin.y + 15
```

here:
```swift
let yPos = (oscPosition == .top || (oscPosition == .floating && sliderFrameInWindow.y + 52 + height >= window!.frame.height)) ?
  sliderFrameInWindow.y - height : sliderFrameInWindow.y + 32
```

and here:
```swift
timePreviewWhenSeek.frame.origin = CGPoint(
  x: round(sender.knobPointPosition() - timePreviewWhenSeek.frame.width / 2),
  y: playSlider.frame.origin.y + 16)
```

This changes the vertical location of the time and thumbnail previews to be relative to the top of the progress bar slider. This prevents the time from intruding into the slider.

The changes in the PR will also update the method `PlaySliderCell.awakeFromNib` to set the slider `controlSize` to `small` for macOS 11+. This causes the slider to be 19 points high under both Catalina and Monterey.

With these changes there is still a problem. As can be seen in this screenshot the top of the time preview is clipped:

![issue-3911-clipped](https://user-images.githubusercontent.com/86170219/185295155-d4c9ae4a-9804-4291-93d7-2c057b98a988.jpg)


This seems to be another confirmation that the OSC was designed for a progress bar that was 15 points high.

To address the clipping of the time the changes in the PR increase the bottom and top layout constraints around the slider from 4.5 to 6. This compresses the slider to be 16 points high on both Catalina and Monterey. This stops the clipping of the time:
![issue-3911-fix](https://user-images.githubusercontent.com/86170219/185295237-efcf84f9-794a-4dda-9f10-e354d9f4029c.jpg)


And of course the green background was only for debugging and is not included in the changes.

**Comment by uiryuu** [comment]
Tested on macOS 10.15, seems fine.

**Comment by low-batt** [comment]
Fixed in IINA 1.3.1.
